processor/transformprocessor/internal/metrics/functions.go

```diff
@@ -15,10 +15,83 @@
 package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/metrics"
 
 import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
 	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/common"
 )
 
+// registry is a map of names to functions for metrics pipelines
+var registry = map[string]interface{}{
+	"convert_sum_to_gauge": convertSumToGauge,
+	"convert_gauge_to_sum": convertGaugeToSum,
+}
+
+func init() {
+	// Init metrics registry with default functions common to all signals
+	for k, v := range common.DefaultFunctions() {
+		registry[k] = v
+	}
+}
+
 func DefaultFunctions() map[string]interface{} {
-	// No metric-only functions yet.
-	return common.DefaultFunctions()
+	return registry
+}
+
+func convertSumToGauge() (common.ExprFunc, error) {
+	return func(ctx common.TransformContext) interface{} {
+		mtc, ok := ctx.(metricTransformContext)
+		if !ok {
+			return nil
+		}
+
+		metric := mtc.GetMetric()
+		if metric.DataType() != pmetric.MetricDataTypeSum {
+			return nil
+		}
+
+		dps := metric.Sum().DataPoints()
+
+		metric.SetDataType(pmetric.MetricDataTypeGauge)
+		// Setting the data type removed all the data points, so we must copy them back to the metric.
+		dps.CopyTo(metric.Gauge().DataPoints())
```

Comment:
From a pdata perspective this will get a metric with the correct data type and leave all the NumberDataPoints intact, but is that all it takes to convert a sum to a gauge?  I want to make sure this is a semantically meaningful conversion.

Reply:
I'll take a close look at this, I see what you mean.

Reply:
Alright, I took a read of https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#metric-points

Sums themselves have the two extra fields, which we take into account by having the user specify these fields when converting from gauge.

For the datapoints, the main semantic difference is the timestamps, but I think this is actually "best-effort" conversion as-is.

For a sum, StartTimestamp -> Timestamp give the range of time the sum is computed over.

For a gauge, StartTimestamp is the earliest point the gauge could be recorded, and Timestamp is when the dp was sampled.

I think these timestamps transfer quite well.

* StartTimestamp in a sum being the start of when collection could occur makes sense. You need to know the value at that point in time in order to compute the value of the sum over the whole time range. Semantically, it seems to make sense.

* Timestamp in a sum could technically be any time after start and before "now" (typically, it's the time when the metric was scraped, but it doesn't need to be). Timestamp would represent the end range of the sum; There would necessarily need to be a sample taken at this point in time in order for the sum to be know, so it makes sense to convert this Timestamp to the gauge Timestamp. This one is definitely the most confusing case, but it seems to me like the conversion makes sense.

So, I do think this is a semantically meaningful conversion, even if most of that conversion is basically a no-op.

Reply:
Yeah, I think the timestamps are logically connected enough that conversion both ways is OK.

Reply:
@open-telemetry/collector-contrib-approvers would also love input from people more versed on otel metrics than I.

Reply:
Thinking about this more, this creates wrong semantics for Gauge -> Delta Sum (time windows shouldn't overlap). I don't have a good solution right this second, will need to think about it a bit more.

Reply:
monotonicity is something else to thing about.  A gauge can go up and down, so setting `is_monotonic=true` when it goes up and down is not semantically correct, but the gauge could technically have only increased, so sometimes `is_monotonic=true` is semantically correct.

I think overall these 2 functions require immense knowledge about the metrics being converted.  They are definitely "Use at your own risk" types of functions.  Should probably add a disclaimer in the readme.

Reply:
Yeah, you're right. I think the purpose of these functions is pretty much the source is mis-reporting the metrics actual semantic type. Converting from Sum->Gauge or vice-versa is already breaking the semantics, technically. 

I added a disclaimer to the README for the functions. I think we shouldn't try to be too clever with the conversion, 'cuz these are really "I know what I'm doing" functions.

Reply:
Ya I agree.

Reply:
I don't think we should have functions that are there just for the purpose of fixing something broken down the pipeline and that are not semantically meaningful. Does this have another use case besides fixing a type that was misreported?

Reply:
I see, thanks for the context, I was on vacation last week and didn't make it to the meeting. If this was decided on the SIG I don't think I should block this of course 😄